Source/Core/DolphinQt2/Translation.cpp

```diff
@@ -0,0 +1,239 @@
+    if (version_major > 1)
+    {
+      ERROR_LOG(COMMON, "MO file '%s' has unsupported version number %i", filename.c_str(),
+                version_major);
+      m_data = {};
+      return;
+    }
+
+    m_number_of_strings = ReadU32(&m_data[8]);
+    m_offset_original_table = ReadU32(&m_data[12]);
+    m_offset_translation_table = ReadU32(&m_data[16]);
+  }
+
+  u32 GetNumberOfStrings() const { return m_number_of_strings; }
+  const char* Translate(const char* original_string) const
+  {
+    auto iter = std::lower_bound(
+        MoIterator(m_data.data(), m_offset_original_table),
+        MoIterator(m_data.data(), m_offset_original_table, m_number_of_strings), original_string,
+        [](const char* a, const char* b) { return strcmp(a, b) < 0; });
+
+    if (strcmp(*iter, original_string) != 0)
+      return original_string;
+
+    u32 offset = ReadU32(&m_data[m_offset_translation_table + iter.GetIndex() * 8 + 4]);
+    return &m_data[offset];
+  }
+
+private:
+  std::vector<char> m_data;
+  u32 m_number_of_strings = 0;
+  u32 m_offset_original_table = 0;
+  u32 m_offset_translation_table = 0;
+};
+
+class MoTranslator : public QTranslator
+{
+public:
+  using QTranslator::QTranslator;
+
+  bool isEmpty() const override { return m_mo_file.GetNumberOfStrings() == 0; }
+  bool load(const std::string& filename)
+  {
+    m_mo_file = MoFile(filename);
+    return !isEmpty();
+  }
+
+  QString translate(const char* context, const char* source_text,
+                    const char* disambiguation = Q_NULLPTR, int n = -1) const override
+  {
+    return QString::fromUtf8(m_mo_file.Translate(source_text));
+  }
+
+private:
+  MoFile m_mo_file;
+};
+
+static bool TryInstallTranslator(const QString& exact_language_code)
+{
+  QStringList possible_language_codes;
+  possible_language_codes << exact_language_code;
+
+  // Qt likes to separate language, script, and country by hyphen, but on disk they're separated by
+  // underscores.
+  possible_language_codes.replaceInStrings(QStringLiteral("-"), QStringLiteral("_"));
+
+  // Try successively dropping subtags (like the stock QTranslator, and as specified by RFC 4647
+  // "Matching of Language Tags").
+  // Example: fr_Latn_CA -> fr_Latn -> fr
+  for (auto lang : QStringList(possible_language_codes))
+  {
+    while (lang.contains(QLatin1Char('_')))
+    {
+      lang = lang.left(lang.lastIndexOf(QLatin1Char('_')));
+      possible_language_codes << lang;
+    }
+  }
+
+  // On macOS, Chinese (Simplified) and Chinese (Traditional) are represented as zh-Hans and
+  // zh-Hant, but on Linux they're represented as zh-CN and zh-TW. Arguably, Qt should probably
+  // include the script subtags on Linux, but it doesn't.
+  if (possible_language_codes.contains(QStringLiteral("zh_Hans")))
+    possible_language_codes << QStringLiteral("zh_CN");
+  if (possible_language_codes.contains(QStringLiteral("zh_Hant")))
+    possible_language_codes << QStringLiteral("zh_TW");
```

Comment:
This and the above code should likely be placed into its own `FindPossibleLanguageCodes()`. Since this is all functionally separate from attempting to actually install the translator.

Reply:
Yeah, good call.